src/databricks/labs/ucx/assessment/dashboards.py

```diff
@@ -0,0 +1,370 @@
+        except DatabricksError as e:
+            logger.warning(f"Cannot get Redash dashboard: {dashboard_id}", exc_info=e)
+            return None
+
+    def _try_fetch(self) -> Iterable[Dashboard]:
+        for row in self._fetch(f"SELECT * FROM {escape_sql_identifier(self.full_name)}"):
+            yield Dashboard(*row)
+
+    def list_queries(self, dashboard: Dashboard | None = None) -> Iterator[Query]:
+        """List queries.
+
+        Args:
+            dashboard (DashboardType | None) : List queries for dashboard. If None, list all queries.
+                Defaults to None.
+
+        Note:
+            This public method does not adhere to the common crawler layout, still, it is implemented to avoid/postpone
+            another crawler for the queries by retrieving the queries every time they are requested.
+        """
+        for query in self._list_legacy_queries(dashboard):
+            yield Query.from_legacy_query(query)
+
+    def _list_legacy_queries(self, dashboard: Dashboard | None = None) -> Iterator[LegacyQuery]:
+        """List legacy queries.
+
+        Args:
+            dashboard (DashboardType | None) : List queries for dashboard. If None, list all queries.
+                Defaults to None.
+
+        Note:
+            This public method does not adhere to the common crawler layout, still, it is implemented to avoid/postpone
+            another crawler for the queries by retrieving the queries every time they are requested.
+        """
+        if dashboard:
+            queries_iterator = self._list_legacy_queries_from_dashboard(dashboard)
+        else:
+            queries_iterator = self._list_all_legacy_queries()
+        # Redash APIs are very slow to paginate, especially for large number of dashboards, so we limit the listing
+        # to a small number of items in debug mode for the assessment workflow just to complete.
+        counter = itertools.count()
+        while self._debug_listing_upper_limit is None or self._debug_listing_upper_limit > next(counter):
+            try:
+                yield next(queries_iterator)
+            except StopIteration:
+                break
+
+    def _list_all_legacy_queries(self) -> Iterator[LegacyQuery]:
+        """List all queries."""
+        if self._include_query_ids:
+            yield from self._get_legacy_queries(*self._include_query_ids)
+        else:
+            try:
+                yield from self._ws.queries_legacy.list()  # TODO: Update this to non-legacy query
+            except DatabricksError as e:
+                logger.warning("Cannot list Redash queries", exc_info=e)
+
+    def _list_legacy_queries_from_dashboard(self, dashboard: Dashboard) -> Iterator[LegacyQuery]:
+        """List queries from dashboard."""
+        if self._include_query_ids:
+            query_ids = set(dashboard.query_ids) & set(self._include_query_ids)
+        else:
+            query_ids = set(dashboard.query_ids)
```

Comment:
I normally prefer this for immutable collections. For mutable ones an alternative could be:
```suggestion
        query_ids = set(dashboard.query_ids)
        if self._include_query_ids:
            query_ids &= set(self._include_query_ids)
```